find() should return 'Not found' for out of scope items

capellini opened the pull request:
Currently `find(arrayWithOutOfScopeItem)` ignores out-of-scope items when returning the array of items.  Instead, `find(...)` should return an 'Item not found' error in place of the out of scope items.

Closes #116.

gr2m commented on tests/specs/scoped.js:
```diff
@@ -82,6 +82,33 @@ test('scoped Store .find()', function (t) {
+  })
+
+  .catch(t.fail)
+})
```
test looks good 👍

capellini commented:
@gr2m -- I noticed that there was one line that wasn't covered under any tests, but since the use case didn't pertain directly to this change, I was going to save it for another PR.  Do you agree, or would you like me to add the test to this PR?

capellini commented on lib/scoped/find.js:
```diff
@@ -7,7 +7,9 @@ function find (type, api, objectsOrIds) {
     if (Array.isArray(storedObjects)) {
-      var typedObjects = storedObjects.filter(typeFilter.bind(null, type))
+      var typedObjects = storedObjects.map(function (storedObject) {
+        return typeFilter(type, storedObject) ? storedObject : new Error('Item not found')
```
I wasn't sure if 'Item not found' was the correct error to throw here.  Would 'Item out of scope' or 'Incorrect item type' be better?

gr2m replied:
I think best would be to throw a similar error as we do here:
https://github.com/hoodiehq/pouchdb-hoodie-api/blob/master/helpers/find-one.js#L14-L16

Maybe instead of `Object with id "' + id + '" is missing` we can say `Object with type "' + type + '" and id "' + id + '" is missing`

gr2m commented:
> I noticed that there was one line in find.js that wasn't covered under any tests, but since the use case didn't pertain directly to this change, I was going to save it for another PR. Do you agree, or would you like me to add the test to this PR?

As you prefer, both is fine with me :) Thanks for your great help Rob, it motivates me a lot to push things forward myself

capellini commented:
I've made the requested changes to the error object.

> As you prefer, both is fine with me :) 

I'm somewhat obsessive about keeping separate concerns on separate issues, so I'll create a separate PR. 

> Thanks for your great help Rob, it motivates me a lot to push things forward myself

You're most welcome!  Me motivating you is motivating me (and now my eyes have crossed!).

capellini commented:
@gr2m -- I just checked out master and ran a coverage test and it looks like this commit somehow IS the culprit of several lines no longer being covered.  On second thought, I'll add the tests to this PR to make sure that we're still at 100% on all files.

gr2m commented:
okay, thanks!

capellini commented on lib/scoped/find.js:
```diff
@@ -18,8 +23,22 @@ function find (type, api, objectsOrIds) {
+}
+
+function allObjectsAreNotFound (typedObjects) {
+  return typedObjects.every(function (typedObject) {
```
Is the use of `Array.every` ok here?  MDN shows the following compatibility tables:
![screenshot from 2016-08-27 00-52-12](https://cloud.githubusercontent.com/assets/75311/18022605/8b74c3fc-6bf0-11e6-9054-1b5e9fcab465.png)
![screenshot from 2016-08-27 00-52-53](https://cloud.githubusercontent.com/assets/75311/18022618/a0b6f762-6bf0-11e6-88f1-261d67e9d5e7.png)

gr2m replied:
yes, we try to match PouchDB’s browser support: https://pouchdb.com/learn.html#browser_support pretty much everything after IE 10 right now

capellini commented:
@gr2m back to 100% coverage.  Please see above comment re: `Array.every`.  Also, note:

* to keep with the DRY principle, since I reused the `isNotFoundError` function from `update-or-add.js`, I moved the code into a module in the `utils` directory
* I had to change the test for `typedObjects.length === 0`, since we're now passing through 'Not Found' errors.

gr2m commented on lib/scoped/find.js:
```diff
@@ -1,15 +1,20 @@
 
-      if (typedObjects.length === 0) {
+      if (allObjectsAreNotFound(typedObjects)) {
         throw new Error('Items not found')
```
I’d argue that we shouldn’t throw an error here. If I call `store.find()` with an array, it should always resolve with an array, even none of the documents could be found. I’d vote for predictability here

capellini replied:
Agreed

gr2m commented on lib/utils/is-not-found-error.js:
```diff
@@ -0,0 +1,5 @@
+
+function isNotFoundError (obj) {
+  return obj instanceof Error && obj.name === 'Not found'
+}
```
👍

gr2m commented on tests/specs/scoped.js:
```diff
@@ -82,6 +82,58 @@ test('scoped Store .find()', function (t) {
+  var store = new Store('test-db-scoped-find', merge({remote: 'test-db-scoped-find'}, options))
+  var testStore = store('test')
+
+  var testItems = [{ foo: 'bar1' }, { foo: 'bar2' }]
```
I’d suggest that you set the `id` property in the `testItems` array (you can use `id` instead of `foo` if you like). Then you know the IDs already and don’t need to read them out from what `.find()` resolves with

gr2m commented on tests/specs/scoped.js:
```diff
@@ -82,6 +82,58 @@ test('scoped Store .find()', function (t) {
+  testStore.add(testItems)
+
+  .then(function (items) {
+    return testStore.find(items.map(function (item) { return {id: item.id + '0'} }))
```
as we just want to test how `.find()` behaves if none of the times can be found, I’d suggest we do something more simple and explicit, like

``` js
return testStore.find([{id: 'non-existing1'}, {id: 'non-existing2'})
```

capellini commented:
@gr2m - Thank you for the feedback.  Changes incorporated.  It looks like consistently returning an array when an array is passed into `find` also makes the code a little cleaner.  There were some `catch` clauses within the `*OrAdd` functions that are no longer reached, so I removed those lines from `update-or-add.js` and `find-or-add.js`.

And it looks like we no longer need the `isNotFoundError` function within `find.js` anymore, so I moved it back into `update-or-add.js`.

gr2m commented:
fantastic work Rob, the PR looks great now! 

LGTM
ping @hoodiehq/maintainers 👀

Charlotteis commented:
LGTM
